File: Solutions/Number_of_Recent_Calls/Number_of_Recent_Calls.cs
public class RecentCounter {

    Queue<int> q {get; set;} 
    
    public RecentCounter() {
        q = new Queue<int>();
    }
    
    public int Ping(int t) {
        q.Enqueue(t);
        while(q.Peek() < t - 3000) {
            q.Dequeue();
        }
        return q.Count();
    }
}

/**
 * Your RecentCounter object will be instantiated and called as such:
 * RecentCounter obj = new RecentCounter();
 * int param_1 = obj.Ping(t);
 */


//Faster variation
public class RecentCounter {

    Queue<int> queue;
    public RecentCounter() {
        queue = new Queue<int>();
    }
    
    public int Ping(int t) {
        queue.Enqueue(t);
        while(queue.First() < t - 3000)
        {
            queue.Dequeue();
        }
        return queue.Count;
    }
}

//Fastest
public class RecentCounter {
    private readonly Queue<int> pings = new Queue<int>();
    public RecentCounter() {
        
    }
    
    public int Ping(int t) {
        if ( t > 3000)
        {
            var cutOff = t - 3000;
            while(pings.Count > 0 && pings.Peek() < cutOff)
            {
                pings.Dequeue();
            }
        }
        
        pings.Enqueue(t);
        return pings.Count;
    }
}

File: Solutions/Number_of_Recent_Calls/Number_of_Recent_Calls.js
var RecentCounter = function () {
  this.arr = [];
  this.start = 0;
};

/** 
 * @param {number} t
 * @return {number}
 */
RecentCounter.prototype.ping = function (t) {
  this.arr.push(t);
  while (this.arr[this.start] < t - 3000) {
    this.start++;
  }
  return this.arr.length - this.start;
};

//Fastest
var RecentCounter = function () {
  this.stream = [];
};

/** 
 * @param {number} t
 * @return {number}
 */
RecentCounter.prototype.ping = function (t) {
  this.stream.push(t);
  while (this.stream[0] < t - 3000) this.stream.shift();
  return this.stream.length;
};

/** 
 * Your RecentCounter object will be instantiated and called as such:
 * var obj = new RecentCounter()
 * var param_1 = obj.ping(t)
 */
